esy-native-install-npm-release/EsyNativeInstallNpmRelease.re

```diff
@@ -0,0 +1,213 @@
+    | Rewrite(storePath) =>
+      EsyBuild.BuildSandbox.importBuild(storePath, filePath)
+    | NoRewrite(storePath) =>
+      let buildId =
+        Str.global_replace(
+          Str.regexp(".tar.gz$"),
+          "",
+          Fpath.basename(filePath),
+        );
+      let storeStagePath = Path.(storePath / storeStageTree);
+      let buildStagePath = Path.(storeStagePath / buildId);
+      let buildFinalPath = Path.(storePath / storeInstallTree / buildId);
+
+      let%bind () = Fs.createDir(buildStagePath);
+      let%bind () = Tarball.unpack(filePath, ~dst=storeStagePath);
+      Fs.rename(~src=buildStagePath, buildFinalPath);
+    }
+  );
+};
+
+let main = rewritePrefix => {
+  let storePath =
+    switch (rewritePrefix) {
+    | Rewrite(path)
+    | NoRewrite(path) => path
+    };
+
+  let check = () => {
+    let%lwt buildFound = Fs.exists(releaseExportPath);
+    switch (buildFound) {
+    | Ok(true) =>
+      switch (rewritePrefix) {
+      | NoRewrite(_) => Lwt.return(Ok())
+      | Rewrite(storePath) =>
+        let%lwt storeFound = Fs.exists(storePath);
+        Lwt.return(
+          switch (storeFound) {
+          | Ok(true) => Error(`ReleaseAlreadyInstalled)
+          | Ok(false) => Ok()
+          | Error(err) => Error(`EsyLibError(err))
+          },
+        );
+      }
+    | Ok(false) => Lwt.return(Error(`NoBuildFound))
+    | Error(err) => Lwt.return(Error(`EsyLibError(err)))
+    };
+  };
+  let initStore = () => {
+    Lwt_result.(
+      Fs.createDir(storePath)
+      >>= (
+        _ =>
+          RunAsync.List.waitAll([
+            Fs.createDir(Path.(storePath / storeBuildTree)),
+            Fs.createDir(Path.(storePath / storeInstallTree)),
+            Fs.createDir(Path.(storePath / storeStageTree)),
+          ])
+      )
+      |> Lwt_result.map_err(err => `EsyLibError(err))
+    );
+  };
+  let doImport = () => {
+    open RunAsync.Syntax;
+    let importBuilds = () => {
+      open RunAsync.Syntax;
+      let%bind files = Fs.listDir(releaseExportPath);
+
+      RunAsync.List.mapAndJoin(
+        ~f=file => importBuild(file, rewritePrefix),
+        files |> List.map(~f=Path.addSeg(releaseExportPath)),
+      );
+    };
+    let rewriteBinWrappers = () =>
+      switch (rewritePrefix) {
+      | NoRewrite(_) => Lwt_result.return()
+      | Rewrite(storePath) =>
+        let%bind prevStorePath =
+          Fs.readFile(Path.(releaseBinPath / "_storePath"));
+        RewritePrefix.rewritePrefix(
+          ~origPrefix=Path.v(prevStorePath),
+          ~destPrefix=storePath,
+          releaseBinPath,
+        );
+      };
+    let%bind _ = importBuilds();
+    rewriteBinWrappers();
+  };
+  Lwt_result.(
+    check()
+    >>= initStore
+    >>= (_ => doImport() |> Lwt_result.map_err(err => `EsyLibError(err)))
+    >>= (
+      _ => {
+        switch (System.Platform.host, System.Arch.host) {
+        | (Darwin, Arm64) =>
+          print_endline("Detected macOS arm64. Signing binaries...");
```

Comment:
I saw it after that you're doing it differently here:

https://github.com/esy/esy/blob/560642c2297bb01a9e740840424ba7847a978ce6/esy-build-package/Build.re#L372

Which one is best ?

Reply:
I completely missed `EsyLib.System` - thanks!

`EsyLib.System` internally runs uname too. Teeny tiny nit - it runs `uname` twice. Definitely, doesnt justify reinventing it again.

+1 to your approach.